### PR TITLE
BLD: Refactor BLAS/LAPACK wrapper infrastructure

### PR DESCRIPTION
**Pull request by thalassemia**
#### Reference issue
Based on the discussion in gh-19816.

#### What does this implement/fix?
I refactored `scipy/linalg/_generate_pyx.py` for better readability and flow. To minimize code duplication in gh-19816, I moved the code shared between this file and that PR into a separate file (`scipy/_build_utils/_wrappers_common.py`) for both to import.

Before this PR, `_generate_pyx.py` generated Fortran subroutine wrappers over complex-valued BLAS/LAPACK functions (see gh-19855). This PR cuts that logic (and Fortran) out by adding equivalent C wrappers to `scipy/_build_utils/src/wrap_g77_abi.c` and `wrap_dummy_g77.abi.c`.

An additional motivation for these hard-coded C wrappers comes from `scipy.linalg._fblas` and `_flapack`, which both currently rely on F2PY-generated Fortran subroutine wrappers. Specifically, in generating these wrappers, F2PY also [adds some incorrect code](https://github.com/scipy/scipy/pull/19816#issuecomment-1970338432) that will cause problems down the line.

To avoid this, I refactored the complex-valued functions in `scipy/linalg/fblas_l1.pyf.src` into subroutines. This stops F2PY from generating both the problematic code and the Fortran wrappers. I replaced these wrappers with my newly added C wrappers by editing the `fortranname` statements for the relevant functions.

While I was at it, I added `intent(c)` statements to the other functions in `fblas_l1.pyf.src` and `flapack_other.pyf.src`, modifying `callstatement` and `callprotoargument` appropriately. This eliminated Fortran from the code generated by F2PY, leaving pure C.

### CODE REVIEW

**Review by h-vetinari — COMMENTED** [review]
Thanks for breaking this out! :)

I'm not familiar with the code in the `*.pyf.src` files, so it would be good if someone else could check that (though it looks sensible to me). Other than that, I'm happy with this, aside from some minor nits.

File: scipy/linalg/_generate_pyx.py
Comment (thalassemia) [reply scipy/linalg/_generate_pyx.py]: The line breaks are a holdover from gh-19816, where `generate_decl_pyx` takes an additional `accelerate=accelerate` kwarg and causes the list comprehension to exceed the line limit.

File: scipy/linalg/meson.build
Comment (thalassemia) [reply scipy/linalg/meson.build:189]: I added a new commit after the first one that explicitly replaces the Fortran wrappers in `_generate_pyx.py` with the C wrappers added in the first commit. This new second commit also includes all the necessary changes to `meson.build` to make this happen.

**Comment by tylerjereddy** [comment]
I'd like to branch for SciPy `1.13.0` around Sunday (17th), let me know if this makes sense to have merged in by then or if we should bump the milestone.

**Comment by thalassemia** [comment]
@tylerjereddy Personally, I think this is ready to go, though it would be nice to have one more reviewer.

**Comment by h-vetinari** [comment]
Both this and https://github.com/scipy/scipy/pull/19816 should make it into 1.13. I'm happy with this PR, but I've been mainly waiting if others (@rgommers @ilayn @larsoner) want to comment as well.

**Comment by h-vetinari** [comment]
@thalassemia, could you please rebase on main? I don't want to squash this PR as the commits are very distinct, but I also don't want to add spurious merge commits.

**Comment by rgommers** [comment]
I'll try to have a look at this tomorrow.

**Review by h-vetinari — APPROVED** [review]
Thank you for breaking this out of #19816! 

The changes LGTM (modulo those in `*.pyf.src`, which I don't feel qualified to review).

@rgommers This PR is a consequence of a [review](https://github.com/scipy/scipy/pull/19816#pullrequestreview-1915673478) I had in the other PR (building upon yet another previous [review](https://github.com/scipy/scipy/pull/19816#discussion_r1446879993)); here we refactor the existing infrastructure based on a setup that allows easy extension for the 13.3+ enablement, without duplicating the things found in `_wrappers_common.py`.

That said, this is still +/- the first complete iteration that @thalassemia and I came up with, so it's possible you'd like to change some things.

PS. Here's the `_cpointer` [comment](https://github.com/scipy/scipy/pull/19816#issuecomment-1970338432) that https://github.com/scipy/scipy/pull/20232/commits/12ac2c452929c833ea0f0c18df5999b5b2443aaa refers to.

**Comment by ilayn** [comment]
I can't spare time  from work too much these days. Hence did not look very deep into things. However I can't see how they can cause any issues either hence looks good to me. 

With little time I have, testing ARPACK/PROPACK rewrite anyways, so soon we can remove these dummy wrappers completely. But probably will happen in 1.14 the earliest.

File: scipy/linalg/meson.build
Comment (rgommers) [inline scipy/linalg/meson.build:22]: This I wasn't 100% sure about; should be fine but in case it gives an issue later we can avoid the problem by importing `_wrappers_common` with `importlib` like so: https://github.com/numpy/numpy/blob/64676baf423c3e1c53b37ae413c3ca6aaecc642a/numpy/_core/code_generators/genapi.py#L27-L32

**Review by rgommers — APPROVED** [review]
This looks quite good, thanks a lot @thalassemia. And thanks for the reviews @h-vetinari.

I agree that this looks ready to go. I did some local testing, including with building scikit-learn on top (because our Cython API is very poorly tested by our own test suite). All looked good, not a single build warning or test failure. And changing to C wrappers will likely have multiple benefits down the line.

**Pull request merged**
